pkg/prowgen/podspec.go

```diff
@@ -0,0 +1,502 @@
+	"reflect"
+	"sort"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	cioperatorapi "github.com/openshift/ci-tools/pkg/api"
+)
+
+var defaultPodSpec = corev1.PodSpec{
+	ServiceAccountName: "ci-operator",
+	Containers: []corev1.Container{
+		{
+			Args: []string{
+				"--image-import-pull-secret=/etc/pull-secret/.dockerconfigjson",
+				"--gcs-upload-secret=/secrets/gcs/service-account.json",
+				"--report-credentials-file=/etc/report/credentials",
+			},
+			Command:         []string{"ci-operator"},
+			Image:           "ci-operator:latest",
+			ImagePullPolicy: corev1.PullAlways,
+			Resources: corev1.ResourceRequirements{
+				Requests: corev1.ResourceList{"cpu": *resource.NewMilliQuantity(10, resource.DecimalSI)},
+			},
+			VolumeMounts: []corev1.VolumeMount{
+				{
+					Name:      "pull-secret",
+					MountPath: "/etc/pull-secret",
+					ReadOnly:  true,
+				},
+				{
+					Name:      "result-aggregator",
+					MountPath: "/etc/report",
+					ReadOnly:  true,
+				},
+				{
+					Name:      "gcs-credentials",
+					MountPath: cioperatorapi.GCSUploadCredentialsSecretMountPath,
+					ReadOnly:  true,
+				},
+			},
+		},
+	},
+	Volumes: []corev1.Volume{
+		{
+			Name: "pull-secret",
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{SecretName: "registry-pull-credentials"},
+			},
+		},
+		{
+			Name: "result-aggregator",
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{SecretName: "result-aggregator"},
+			},
+		},
+	},
+}
+
+type ciOperatorPodSpecGenerator struct {
+	mutators    []PodSpecMutator
+	buildErrors []error
+}
+
+// PodSpecMutator is a mutation function operating over a PodSpec pointer.
+// The changes performed on the PodSpec must not depend on any previous
+// state of the PodSpec, and must resolve any conflicts it encounters.
+// The mutator function must also correctly handle being called multiple calls
+type PodSpecMutator func(spec *corev1.PodSpec) error
+
+// CiOperatorPodSpecGenerator is a builder-pattern for building PodSpecs that
+// run ci-operator for various purposes, ensuring the consistency of volumes,
+// mounts, parameters and other content.
+type CiOperatorPodSpecGenerator interface {
+	// Add adds one or more mutations to be performed to build the final PodSpec
+	Add(mutators ...PodSpecMutator) CiOperatorPodSpecGenerator
+	// Build returns the PodSpec built by taking the default PodSpec, applying
+	// all added mutators and sorting several list fields by their keys.
+	Build() (*corev1.PodSpec, error)
+}
+
+// NewCiOperatorPodSpecGenerator returns a new CiOperatorPodSpecGenerator instance
+func NewCiOperatorPodSpecGenerator() CiOperatorPodSpecGenerator {
+	return &ciOperatorPodSpecGenerator{}
+}
+
+func aggregateMutator(mutators ...PodSpecMutator) PodSpecMutator {
+	return func(spec *corev1.PodSpec) error {
+		var errs []error
+		for _, f := range mutators {
```

Comment:
There is a chance here that one mutator can revert changes the happened through another mutator. Do we care?